# HG changeset 9731 patch
# User kaf24@firebug.cl.cam.ac.uk
# Node ID de0e196e1a045974cb242dff2dbdf75070f400ce
# Parent  d83290c8ad8dd56e445dc085d185f8c8f8975860
Subject: Allow ballon driver to pick up extra pages
Initialise pages outside initial allocation so that they are picked up by
the balloon driver.
From: Jan Beulich <jbeulich@novell.com>
References: 184487
Signed-off-by: Keir Fraser <keir@xensource.com>
xen-unstable changeset:   10310:4df81d20a9a1a8dfbe47d1b5697122559d7b6cc9
xen-unstable date:        Fri Jun  9 16:18:40 2006 +0100

Plus removal of the bogus totalram_pages increment.

Index: head-2006-06-07/arch/i386/mm/init-xen.c
===================================================================
--- head-2006-06-07.orig/arch/i386/mm/init-xen.c	2006-06-12 10:04:36.000000000 +0200
+++ head-2006-06-07/arch/i386/mm/init-xen.c	2006-06-12 10:07:09.000000000 +0200
@@ -666,7 +666,6 @@ void __init mem_init(void)
 	for (pfn = xen_start_info->nr_pages; pfn < max_low_pfn; pfn++) {
 		ClearPageReserved(&mem_map[pfn]);
 		set_page_count(&mem_map[pfn], 1);
-		totalram_pages++;
 	}
 
 	reservedpages = 0;
Index: head-2006-06-07/arch/x86_64/mm/init-xen.c
===================================================================
--- head-2006-06-07.orig/arch/x86_64/mm/init-xen.c	2006-06-12 10:04:36.000000000 +0200
+++ head-2006-06-07/arch/x86_64/mm/init-xen.c	2006-06-12 10:07:09.000000000 +0200
@@ -876,6 +876,7 @@ static struct kcore_list kcore_mem, kcor
 void __init mem_init(void)
 {
 	long codesize, reservedpages, datasize, initsize;
+	unsigned long pfn;
 
 	contiguous_bitmap = alloc_bootmem_low_pages(
 		(end_pfn + 2*BITS_PER_LONG) >> 3);
@@ -904,6 +905,11 @@ void __init mem_init(void)
 #else
 	totalram_pages = free_all_bootmem();
 #endif
+	/* XEN: init and count pages outside initial allocation. */
+	for (pfn = xen_start_info->nr_pages; pfn < max_pfn; pfn++) {
+		ClearPageReserved(&mem_map[pfn]);
+		set_page_count(&mem_map[pfn], 1);
+	}
 	reservedpages = end_pfn - totalram_pages - e820_hole_size(0, end_pfn);
 
 	after_bootmem = 1;
